### ISSUE TITLE
Bootloader Failing to Flash RCT on Mac

### ISSUE THREAD

**Issue by stieg** [autosportlabs/ASL_F4_bootloader#20]
Stack coming, but it seems to have a communication error during the flashing process.  Something may be interfering... still digging.

Verion 0.3.0

**Comment by stieg** [comment]
Looks like something in the RCT bootloader is dying.  Here is my flash operation:

```
-- flash_command --
-offset=134258176
-data_len=64
-data=[1543176486, 15914101, 373926425, 589496470, 541523447, 3959362633, 99689730, 1596129430, 541523447, 3825147974, 273230406, 273415927, 2281570800, 3523019040, 4273799359, 1773732360, 1874395656, 381157896, 1958281736, 1102643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2277048840, 2428043784, 2511929864, 1907359752, 1890911302, 32513787, 105259078, 4210535677, 541472329, 4210527485, 894246982, 894040823, 2281513030, 2532849703, 19084105, 4210492157, 541469254, 19083593, 4210505981, 709230627, 541470537, 4210504445, 541458721, 4210528253, 541469769, 4210519293, 116778853, 541497080, 1395065123, 675937015, 3858505798, 99687426, 19080777, 4210498301, 541459953, 989987107, 592050935, 335349233, 1510080547]


316
-- flash_response --
-error=0


-- flash_command --
-offset=134258432
-data_len=64
-data=[541466953, 4210494717, 541458721, 4210518525, 541466185, 4210509565, 541497080, 3559325987, 457833207, 3254526022, 99720450, 19077449, 4210554620, 99729666, 2302022, 373947127, 4026277958, 2226935, 1291657286, 2226935, 1224548422, 2226935, 1157431840, 1891434687, 2662924808, 61211144, 1773732360, 1874395656, 381157896, 1958281736, 1102643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2511929864, 2277048840, 2428043784, 951387206, 32541434, 88481862, 4210497789, 541479753, 4210555132, 3052979220, 2097648, 620279288, 3826548914, 32514040, 3052930069, 2159149552, 452507128, 605388978, 32511480, 3052946453, 2159149552, 284734968, 1679130802, 32508920, 3052962837, 2159149552, 116962808, 2752872626]


310
Retrying
Traceback (most recent call last):
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/bin/asl_f4_loader", line 9, in <module>
    load_entry_point('asl-f4-loader==0.3.0', 'console_scripts', 'asl_f4_loader')()
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py", line 253, in main
    fu.update_firmware(options.input_filename, port, verbose_mode)
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py", line 174, in update_firmware
    raise Exception("Error flashing device, no response!")
Exception: Error flashing device, no response!
(mac_venv) Ans-MacBook-Air:stieg anphan$ asl_f4_loader -v VERBOSE -f ./RaceCapture_2.10.0_RC1-153-ge746a10/RaceCapture_2.10.0_RC1-153-ge746a10.ihex
Place your device into bootloader mode and then press any key
Searching for an active bootloader
Unable to locate an active bootloader
Usage: asl_f4_loader [options]

Options:
  -h, --help            show this help message and exit
  -f INPUT_FILENAME, --filename=INPUT_FILENAME
                        Filename of .ihex file to process
  -p SER_PORT, --port=SER_PORT
                        (optional)path to serial port we're talking on ex:
                        /dev/ttyACM0
  -v VERBOSE, --verbose=VERBOSE
                        turn on verbose debugging
```

Notice how the bootloader does not respond on the second attempt.  You have to power cycle the hardware to get it to be responsive again.  When I try the operation again after a power cycle I see the following (did 3 times to ensure consistency):

```
-- flash_command --
-offset=134258176
-data_len=64
-data=[1543176486, 15914101, 373926425, 589496470, 541523447, 3959362633, 99689730, 1596129430, 541523447, 3825147974, 273230406, 273415927, 2281570800, 3523019040, 4273799359, 1773732360, 1874395656, 381157896, 1958281736, 1102643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2277048840, 2428043784, 2511929864, 1907359752, 1890911302, 32513787, 105259078, 4210535677, 541472329, 4210527485, 894246982, 894040823, 2281513030, 2532849703, 19084105, 4210492157, 541469254, 19083593, 4210505981, 709230627, 541470537, 4210504445, 541458721, 4210528253, 541469769, 4210519293, 116778853, 541497080, 1395065123, 675937015, 3858505798, 99687426, 19080777, 4210498301, 541459953, 989987107, 592050935, 335349233, 1510080547]


316
-- flash_response --
-error=0


-- flash_command --
-offset=134258432
-data_len=64
-data=[541466953, 4210494717, 541458721, 4210518525, 541466185, 4210509565, 541497080, 3559325987, 457833207, 3254526022, 99720450, 19077449, 4210554620, 99729666, 2302022, 373947127, 4026277958, 2226935, 1291657286, 2226935, 1224548422, 2226935, 1157431840, 1891434687, 2662924808, 61211144, 1773732360, 1874395656, 381157896, 1958281736, 1102643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2511929864, 2277048840, 2428043784, 951387206, 32541434, 88481862, 4210497789, 541479753, 4210555132, 3052979220, 2097648, 620279288, 3826548914, 32514040, 3052930069, 2159149552, 452507128, 605388978, 32511480, 3052946453, 2159149552, 284734968, 1679130802, 32508920, 3052962837, 2159149552, 116962808, 2752872626]


310
Retrying
Traceback (most recent call last):
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/bin/asl_f4_loader", line 9, in <module>
    load_entry_point('asl-f4-loader==0.3.0', 'console_scripts', 'asl_f4_loader')()
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py", line 253, in main
    fu.update_firmware(options.input_filename, port, verbose_mode)
  File "/Users/anphan/stieg/RaceCapture_App/mac_venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py", line 174, in update_firmware
    raise Exception("Error flashing device, no response!")
Exception: Error flashing device, no response!
(mac_venv) Ans-MacBook-Air:stieg anphan$
```

Our bootloader crashes at the same offset every time.  Digging...

**Comment by stieg** [comment]
Here is the same spot on my mac using same bootloader version 0.3.0:

```
-- flash_command --
-offset=134258176
-data_len=64
-data=[1543176486, 15914101, 373926425, 589496470, 541523447, 3959362633, 99689730, 1596129430, 541523447, 3825147974, 273230406, 273415927, 2281570800, 3523019040, 4273799359, 1773732360, 1874395656, 381157896, 1958281736, 11
02643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2277048840, 2428043784, 2511929864, 1907359752, 1890911302, 32513787, 105259078, 4210535677, 541472329, 4210527485, 894246982, 894040823, 2281513030, 253284
9703, 19084105, 4210492157, 541469254, 19083593, 4210505981, 709230627, 541470537, 4210504445, 541458721, 4210528253, 541469769, 4210519293, 116778853, 541497080, 1395065123, 675937015, 3858505798, 99687426, 19080777, 42104983
01, 541459953, 989987107, 592050935, 335349233, 1510080547]


316
-- flash_response --
-error=0


-- flash_command --
-offset=134258432
-data_len=64
-data=[541466953, 4210494717, 541458721, 4210518525, 541466185, 4210509565, 541497080, 3559325987, 457833207, 3254526022, 99720450, 19077449, 4210554620, 99729666, 2302022, 373947127, 4026277958, 2226935, 1291657286, 2226935, 1224548422, 2226935, 1157431840, 1891434687, 2662924808, 61211144, 1773732360, 1874395656, 381157896, 1958281736, 1102643720, 2042167816, 1236861448, 1371079176, 1505296904, 2159608328, 2511929864, 2277048840, 2428043784, 951387206, 32541434, 88481862, 4210497789, 541479753, 4210555132, 3052979220, 2097648, 620279288, 3826548914, 32514040, 3052930069, 2159149552, 452507128, 605388978, 32511480, 3052946453, 2159149552, 284734968, 1679130802, 32508920, 3052962837, 2159149552, 116962808, 2752872626]


310
-- flash_response --
-error=0


-- flash_command --
-offset=134258688
-data_len=64
-data=[32506360, 2159149552, 771243337, 38142243, 541522679, 1241298424, 3290702513, 3052987428, 3283135, 19013702, 19084873, 4210507004, 3052930101, 541465116, 415170850, 843645219, 4210504444, 3052938277, 541458482, 415170850, 776536355, 4210501884, 3052946485, 541465116, 415170850, 709427491, 4210499324, 3052954661, 541458482, 415170850, 642318627, 4210496764, 3052962869, 541465116, 415170850, 575209763, 4210494204, 3052971045, 541458482, 415170850, 508100899, 4210491644, 99980645, 541465673, 4210522364, 541459953, 2348941603, 340392695, 805052486, 99707906, 19075145, 4210501884, 99699714, 2302022, 189397751, 570171462, 2226935, 2130452550, 2226935, 2063343686, 2226935, 1996227104, 951910591, 2797142536]


311
-- flash_response --
-error=0
```

Works just fine.  Binary data matches.

**Comment by stieg** [comment]
Fixed by PR #21
